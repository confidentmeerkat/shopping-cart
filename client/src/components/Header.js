import React from 'react';
import { connect } from 'react-redux';
import { getUser } from '../actions/userActions';
import { Link } from 'react-router-dom';
import AppBar from 'material-ui/AppBar';
import FlatButton from 'material-ui/FlatButton';
import IconMenu from 'material-ui/IconMenu';
import MenuItem from 'material-ui/MenuItem';
import IconButton from 'material-ui/IconButton';
import Person from 'material-ui/svg-icons/social/person';
import Menu from 'material-ui/svg-icons/navigation/menu';
import Logout from 'material-ui/svg-icons/navigation/subdirectory-arrow-left';
import ShoppingCart from 'material-ui/svg-icons/action/shopping-cart.js';
import Input from 'material-ui/svg-icons/action/input.js';
import LoginModal from './LoginModal';
import RegisterModal from './RegisterModal';
import '../styles/Header.css';

const styles = {
  menuBtn: {
    color: '#fff'
  },
  iconMenuBtn: {
    color: '#00BCD4',
    minWidth: '168px',
    textAlign: 'left'
  }
}

class Header extends React.Component {

  state = {
    loginModalOpen: false,
    registerModalOpen: false
  }

  toggleOpen = (targetComponent) => {
    this.setState({ [targetComponent]: !this.state[targetComponent] });
  }

  switchLoginRegister = () => {
    this.setState({ 
      loginModalOpen: !this.state.loginModalOpen,
      registerModalOpen: !this.state.registerModalOpen 
    });
  }

  componentDidMount() {
    this.props.getUser();
  }

  render() {
    return (
      <div className="header">
        <AppBar 
          className="title"
          title="MOBILE SHOP"
          onTitleClick={() => this.props.history.push('/')}
          showMenuIconButton={false}
          zDepth={0}
          iconElementRight={
            this.props.loggedUser ? 
              <div>
                <div className="menu">
                  <FlatButton 
                    style={styles.menuBtn}
                    label="ACCOUNT" 
                    icon={<Person />} 
                    containerElement={<Link to="/account" />}
                  />
                  <FlatButton 
                    style={styles.menuBtn}
                    label="CART" 
                    icon={<ShoppingCart />} 
                    containerElement={<Link to="/cart" />}
                  />
                  <FlatButton 
                    style={styles.menuBtn}
                    label="LOGOUT" 
                    icon={<Logout />} 
                    containerElement={<a href="/auth/logout"/>}
                  />
                </div>
                <div className="icon-menu">
                  <IconMenu
                    iconButtonElement={<IconButton><Menu /></IconButton>}
                    anchorOrigin={{ horizontal: 'right', vertical: 'top' }}
                    targetOrigin={{ horizontal: 'right', vertical: 'top' }}
                    iconStyle={{ color: '#fff' }}
                  >
                    <FlatButton
                      style={styles.iconMenuBtn}
                      label="ACCOUNT"
                      icon={<Person />}
                      containerElement={<Link to="/account" />}
                    /><br />
                    <FlatButton
                      style={styles.iconMenuBtn}
                      label="CART"
                      icon={<ShoppingCart />}
                      containerElement={<Link to="/cart" />}
                    /><br />
                    <FlatButton
                      style={styles.iconMenuBtn}
                      label="LOGOUT"
                      icon={<Logout />}
                      containerElement={<a href="/auth/logout" />}
                    />
                  </IconMenu>
                </div>
              </div> : 
              <FlatButton 
                label="LOGIN" 
                labelPosition="before" 
                icon={<Input/>} 
                onClick={() => this.toggleOpen('loginModalOpen')} 
              />
          }
        />
        <LoginModal 
          isOpen={this.state.loginModalOpen} 
          onRequestClose={() => this.toggleOpen('loginModalOpen')} 
          switch={this.switchLoginRegister} 
        />
        <RegisterModal 
          isOpen={this.state.registerModalOpen} 
          onRequestClose={() => this.toggleOpen('registerModalOpen')} 
          switch={this.switchLoginRegister} 
        />
      </div>
    )
  }
};

const mapStateToProps = (state) => ({
  loggedUser: state.loggedUser,
  cart: state.cart
});

const mapDispatchToProps = (dispatch) => ({
  getUser: () => dispatch(getUser())
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);