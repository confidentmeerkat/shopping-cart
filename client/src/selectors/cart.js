export const selectCart = (state) => state.cart;
export const selectItems = (state) => state.cart.items;