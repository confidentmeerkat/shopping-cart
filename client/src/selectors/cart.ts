import { IState, ICart, ICartProduct } from '@typings/state/index';

export const selectCart = (state: IState): ICart => state.cart;
export const selectItems = (state: IState): ICartProduct[] => state.cart.items;
