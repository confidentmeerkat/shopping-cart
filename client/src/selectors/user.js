export const selectUser = (state) => state.loggedUser.user;