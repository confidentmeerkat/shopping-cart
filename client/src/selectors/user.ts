import { IState, IUser } from '@typings/state/index';

export const selectUser = (state: IState): IUser | null => state.loggedUser.user;
