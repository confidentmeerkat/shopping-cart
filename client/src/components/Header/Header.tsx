import * as React from 'react';
import { Link } from 'react-router-dom';
import AppBar from 'material-ui/AppBar';
import FlatButton from 'material-ui/FlatButton';
import IconMenu from 'material-ui/IconMenu';
import IconButton from 'material-ui/IconButton';
import Person from 'material-ui/svg-icons/social/person';
import Menu from 'material-ui/svg-icons/navigation/menu';
import Logout from 'material-ui/svg-icons/navigation/subdirectory-arrow-left';
import ShoppingCart from '@material-ui/icons/ShoppingCart';
import Input from '@material-ui/icons/Input';
import { RouteComponentProps } from 'react-router-dom';
import { ILoggedUser } from '@typings/state/loggedUser';
import { ICart } from '@typings/state/cart';
import { modal } from '@typings/modal';
import LoginModal from '../LoginModal';
import RegisterModal from '../RegisterModal';
import '@styles/Header.css';

interface Props extends RouteComponentProps {
  loggedUser: ILoggedUser;
  cart: ICart;
  getUser: () => void;
}

const styles = {
  menuBtn: {
    color: '#fff'
  },
  iconMenuBtn: {
    color: '#00BCD4',
    minWidth: '168px',
    textAlign: 'left'
  }
}

class Header extends React.Component<Props> {
  state = {
    activeModal: null
  }

  setActiveModal = (modal: modal) => {
    this.setState({ activeModal: modal });
  }

  componentDidMount() {
    this.props.getUser();
  }

  render() {
    const {
      history,
      loggedUser
    } = this.props;

    return (
      <div className="header">
        <AppBar 
          className="title"
          title="MOBILE SHOP"
          onTitleClick={() => history.push('/')}
          showMenuIconButton={false}
          zDepth={0}
          iconElementRight={
            loggedUser ?
              <div>
                <div className="menu">
                  <FlatButton
                    style={styles.menuBtn}
                    label="ACCOUNT"
                    icon={<Person />}
                    containerElement={<Link to="/account" />}
                  />
                  <FlatButton
                    style={styles.menuBtn}
                    label="CART"
                    icon={<ShoppingCart />}
                    containerElement={<Link to="/cart" />}
                  />
                  <FlatButton
                    style={styles.menuBtn}
                    label="LOGOUT"
                    icon={<Logout />}
                    containerElement={<a href="/auth/logout"/>}
                  />
                </div>
                <div className="icon-menu">
                  <IconMenu
                    iconButtonElement={<IconButton><Menu /></IconButton>}
                    anchorOrigin={{ horizontal: 'right', vertical: 'top' }}
                    targetOrigin={{ horizontal: 'right', vertical: 'top' }}
                    iconStyle={{ color: '#fff' }}
                  >
                    <FlatButton
                      label="ACCOUNT"
                      icon={<Person />}
                      containerElement={<Link to="/account" />}
                    /><br />
                    <FlatButton
                      label="CART"
                      icon={<ShoppingCart />}
                      containerElement={<Link to="/cart" />}
                    /><br />
                    <FlatButton
                      label="LOGOUT"
                      icon={<Logout />}
                      containerElement={<a href="/auth/logout" />}
                    />
                  </IconMenu>
                </div>
              </div> :
              <FlatButton
                label="LOGIN"
                labelPosition="before"
                icon={<Input/>}
                onClick={() => this.setActiveModal('login')}
              />
          }
        />
        <LoginModal
          isOpen={this.state.activeModal === 'login'}
          onRequestClose={() => this.setActiveModal(null)}
          setActiveModal={this.setActiveModal}
        />
        <RegisterModal
          isOpen={this.state.activeModal === 'register'}
          onRequestClose={() => this.setActiveModal('register')}
          setActiveModal={this.setActiveModal}
        />
      </div>
    )
  }
};

export default Header;
