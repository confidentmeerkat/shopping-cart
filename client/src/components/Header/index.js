import { compose } from 'recompose';
import { connect } from 'react-redux';
import { getUser } from '@actions';
import { selectUser } from '@selectors/user';
import { selectCart } from '@selectors/cart';
import Header from './Header';

const mapStateToProps = (state) => ({
  loggedUser: selectUser(state),
  cart: selectCart(state)
});

const actions = { getUser };

export default compose(
  connect(mapStateToProps, actions)
)(Header);
