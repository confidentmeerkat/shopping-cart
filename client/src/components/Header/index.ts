import { compose } from 'recompose';
import { connect } from 'react-redux';
import { getUser } from '@actions/index';
import { selectUser } from '@selectors/user';
import { selectCart } from '@selectors/cart';
import { IState } from '@typings/state/index';
import Header from './Header';

const mapStateToProps = (state: IState) => ({
  loggedUser: selectUser(state),
  cart: selectCart(state)
});

const actions = { getUser };

export default compose(
  connect(mapStateToProps, actions)
)(Header);
