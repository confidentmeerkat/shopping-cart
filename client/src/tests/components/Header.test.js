import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';
import user from '../__mocks__/user';
import cart from '../__mocks__/cart';

describe('<Header />', () => {
  it('renders Header with no user logged correctly', () => {
    const wrapper = shallow(<Header
      loggedUser=""
      cart={{ items: [] }}
      getUser={() => { }}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders Header with user logged correctly', () => {
    const wrapper = shallow(<Header
      loggedUser={user}
      cart={cart}
      getUser={() => { }}
    />);

    expect(wrapper).toMatchSnapshot();
  });
});