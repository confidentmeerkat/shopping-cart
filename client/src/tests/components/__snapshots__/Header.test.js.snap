// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders Header with no user logged correctly 1`] = `
<div
  className="header"
>
  <AppBar
    className="title"
    iconElementRight={
      <FlatButton
        disabled={false}
        fullWidth={false}
        icon={<ActionInput />}
        label="LOGIN"
        labelPosition="before"
        labelStyle={Object {}}
        onClick={[Function]}
        onKeyboardFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchStart={[Function]}
        primary={false}
        secondary={false}
      />
    }
    onTitleClick={[Function]}
    showMenuIconButton={false}
    title="MOBILE SHOP"
    zDepth={0}
  />
  <LoginModal
    isOpen={false}
    onRequestClose={[Function]}
    switch={[Function]}
  />
  <RegisterModal
    isOpen={false}
    onRequestClose={[Function]}
    switch={[Function]}
  />
</div>
`;

exports[`<Header /> renders Header with user logged correctly 1`] = `
<div
  className="header"
>
  <AppBar
    className="title"
    iconElementRight={
      <div>
        <div
          className="menu"
        >
          <FlatButton
            containerElement={
              <Link
                replace={false}
                to="/account"
              />
            }
            disabled={false}
            fullWidth={false}
            icon={<SocialPerson />}
            label="ACCOUNT"
            labelPosition="after"
            labelStyle={Object {}}
            onKeyboardFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onTouchStart={[Function]}
            primary={false}
            secondary={false}
            style={
              Object {
                "color": "#fff",
              }
            }
          />
          <FlatButton
            containerElement={
              <Link
                replace={false}
                to="/cart"
              />
            }
            disabled={false}
            fullWidth={false}
            icon={<ActionShoppingCart />}
            label="CART"
            labelPosition="after"
            labelStyle={Object {}}
            onKeyboardFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onTouchStart={[Function]}
            primary={false}
            secondary={false}
            style={
              Object {
                "color": "#fff",
              }
            }
          />
          <FlatButton
            containerElement={
              <a
                href="/auth/logout"
              />
            }
            disabled={false}
            fullWidth={false}
            icon={<NavigationSubdirectoryArrowLeft />}
            label="LOGOUT"
            labelPosition="after"
            labelStyle={Object {}}
            onKeyboardFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onTouchStart={[Function]}
            primary={false}
            secondary={false}
            style={
              Object {
                "color": "#fff",
              }
            }
          />
        </div>
        <div
          className="icon-menu"
        >
          <IconMenu
            anchorOrigin={
              Object {
                "horizontal": "right",
                "vertical": "top",
              }
            }
            animated={true}
            clickCloseDelay={200}
            iconButtonElement={
              <IconButton
                disableTouchRipple={false}
                disabled={false}
                iconStyle={Object {}}
                tooltipPosition="bottom-center"
                touch={false}
              >
                <NavigationMenu />
              </IconButton>
            }
            iconStyle={
              Object {
                "color": "#fff",
              }
            }
            multiple={false}
            onClick={[Function]}
            onItemClick={[Function]}
            onKeyboardFocus={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onRequestChange={[Function]}
            open={null}
            targetOrigin={
              Object {
                "horizontal": "right",
                "vertical": "top",
              }
            }
            useLayerForClickAway={false}
          >
            <FlatButton
              containerElement={
                <Link
                  replace={false}
                  to="/account"
                />
              }
              disabled={false}
              fullWidth={false}
              icon={<SocialPerson />}
              label="ACCOUNT"
              labelPosition="after"
              labelStyle={Object {}}
              onKeyboardFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onTouchStart={[Function]}
              primary={false}
              secondary={false}
              style={
                Object {
                  "color": "#00BCD4",
                  "minWidth": "168px",
                  "textAlign": "left",
                }
              }
            />
            <br />
            <FlatButton
              containerElement={
                <Link
                  replace={false}
                  to="/cart"
                />
              }
              disabled={false}
              fullWidth={false}
              icon={<ActionShoppingCart />}
              label="CART"
              labelPosition="after"
              labelStyle={Object {}}
              onKeyboardFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onTouchStart={[Function]}
              primary={false}
              secondary={false}
              style={
                Object {
                  "color": "#00BCD4",
                  "minWidth": "168px",
                  "textAlign": "left",
                }
              }
            />
            <br />
            <FlatButton
              containerElement={
                <a
                  href="/auth/logout"
                />
              }
              disabled={false}
              fullWidth={false}
              icon={<NavigationSubdirectoryArrowLeft />}
              label="LOGOUT"
              labelPosition="after"
              labelStyle={Object {}}
              onKeyboardFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onTouchStart={[Function]}
              primary={false}
              secondary={false}
              style={
                Object {
                  "color": "#00BCD4",
                  "minWidth": "168px",
                  "textAlign": "left",
                }
              }
            />
          </IconMenu>
        </div>
      </div>
    }
    onTitleClick={[Function]}
    showMenuIconButton={false}
    title="MOBILE SHOP"
    zDepth={0}
  />
  <LoginModal
    isOpen={false}
    onRequestClose={[Function]}
    switch={[Function]}
  />
  <RegisterModal
    isOpen={false}
    onRequestClose={[Function]}
    switch={[Function]}
  />
</div>
`;
